kable-core/src/androidMain/kotlin/Connection.kt

```diff
@@ -2,125 +2,271 @@ package com.juul.kable
-                val response = it.await()
-                logger.warn {
-                    message = "Discarded response"
-                    detail("response", response.toString())
+    suspend inline fun <reified T : Response> execute(
+        noinline action: BluetoothGatt.() -> Unit,
+    ): T = execute(T::class, action)
+
+    suspend fun <T : Response> execute(
+        type: KClass<T>,
+        action: BluetoothGatt.() -> Unit,
+    ): T {
+        val response = guard.withLock {
+            var executed = false
+            try {
+                withContext(dispatcher) {
+                    gatt.action()
+                    executed = true
                 }
+            } catch (e: CancellationException) {
+                if (executed) {
+                    // Ensure response buffer is received even when calling context is cancelled.
+                    // UNDISPATCHED to ensure we're within the `lock` for the `receive`.
+                    connectionScope.launch(start = UNDISPATCHED) {
+                        val response = callback.onResponse.receive()
+                        logger.debug {
+                            message = "Discarded response to cancelled request"
+                            detail("response", response.toString())
+                        }
+                    }
+                }
+                coroutineContext.ensureActive()
+                throw e.unwrapCancellationException()
             }
-        }
 
-        withContext(dispatcher) {
-            bluetoothGatt.action()
-        }
-        val deferred = scope.async { callback.onResponse.receive() }
-        deferredResponse = deferred
-
-        val response = try {
-            deferred.await()
-        } catch (e: Exception) {
-            // The `async` above is a sibling coroutine to that of the coroutines launched from the
-            // connect process. When (for example) BLE is disabled, it fails the connection scope;
-            // being that `async` is a sibling scope, the `async` above will **cancel** when BLE is
-            // disabled (rather than fail). We actually want to fail this `execute`, so we unwrap
-            // `CancellationException` for the underlying exception that failed the connection scope.
-            // todo: Figure out how to handle cancellation (discard GATT callback response) while running in the calling coroutine context (rather than using passed in scope).
-            when (val unwrapped = e.unwrapCancellationCause()) {
-                is ConnectionLostException -> throw ConnectionLostException(cause = unwrapped)
-                else -> throw unwrapped
+            try {
+                connectionScope.async {
+                    callback.onResponse.receive()
+                }.await()
+            } catch (e: CancellationException) {
+                coroutineContext.ensureActive()
+                throw e.unwrapCancellationException()
```

Comment:
This pattern (repeated a few times but I'll only comment on it the once here) is really confusing. If we catch a `CancellationException` and then call `ensureActive`, it seems like that is guaranteed to re-throw the `CancellationException`, and means we would never reach the following code where we re-throw the `e.unwrapCancellationException()`.

I assume I am misunderstanding something here but this code reads like it could be removed and the behavior would not change. 😕

Reply:
There are two possible sources for cancellation:
1. We're running in `suspend execute(..)` function (line 149)
2. Suspended `Channel.receive()` could cancel/throw

Essentially, cancellation could occur on either the consumer side _or_ the producer side.

The beauty of `ensureActive` is it says we should propagate a `CancellationException` iff calling side cancelled. i.e. If the coroutine context we're executing in was cancelled then let the cancellation occur.

Whereas, if cancellation occurred on the producer side (from the `Channel.receive()` then the consumer side coroutine context would still be active (and `ensureActive` would not throw).

So, we'll only unwrap the `CancellationException` if the cancellation occurred on the producer side, but we'll honor cancellation that a library consumer initiated.

Reply:
Oooh, I see. Makes perfect sense in the broader context, thanks for explaining!